fe/fe-core/src/main/java/com/starrocks/system/TabletComputeNodeMapper.java

```diff
@@ -0,0 +1,187 @@
+        }
+
+        // Returns whether any single compute node has been added to this TabletMap
+        private boolean tracksSomeComputeNode() {
+            return !tabletToComputeNodeId.get(ARBITRARY_FAKE_TABLET, 1).isEmpty();
+        }
+    }
+
+    private final Map<String, TabletMap> resourceIsolationGroupToTabletMapping;
+    private final ReentrantReadWriteLock stateLock = new ReentrantReadWriteLock();
+    private final Lock readLock = stateLock.readLock();
+    private final Lock writeLock = stateLock.writeLock();
+
+    public TabletComputeNodeMapper() {
+        resourceIsolationGroupToTabletMapping = new HashMap<>();
+    }
+
+    @TestOnly
+    public void clear() {
+        resourceIsolationGroupToTabletMapping.clear();
+    }
+
+    public int numResourceIsolationGroups() {
+        return resourceIsolationGroupToTabletMapping.size();
+    }
+
+    public void addComputeNode(Long computeNodeId, String resourceIsolationGroup) {
+        writeLock.lock();
+        try {
+            addComputeNodeUnsynchronized(computeNodeId, resourceIsolationGroup);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    private void maybeInitResourceIsolationGroup(String resourceIsolationGroup) {
+        if (resourceIsolationGroup == null) {
+            resourceIsolationGroup = DEFAULT_RESOURCE_ISOLATION_GROUP_ID;
+        }
+        if (!resourceIsolationGroupToTabletMapping.containsKey(resourceIsolationGroup)) {
+            resourceIsolationGroupToTabletMapping.put(resourceIsolationGroup, new TabletMap());
+        }
+    }
+
+    private void addComputeNodeUnsynchronized(long computeNodeId, String resourceIsolationGroup) {
+        maybeInitResourceIsolationGroup(resourceIsolationGroup);
+        TabletMap map = resourceIsolationGroupToTabletMapping.get(resourceIsolationGroup);
+        map.tabletToComputeNodeId.addNode(computeNodeId);
+    }
+
+    // This will succeed even if the resource isolation group is not being tracked.
+    public void removeComputeNode(Long computeNodeId, String resourceIsolationGroup) {
+        writeLock.lock();
+        try {
+            removeComputeNodeUnsynchronized(computeNodeId, resourceIsolationGroup);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    private void removeComputeNodeUnsynchronized(Long computeNodeId, String resourceIsolationGroup) {
+        TabletMap map = resourceIsolationGroupToTabletMapping.get(resourceIsolationGroup);
+        if (map == null) {
+            return;
+        }
+        map.tabletToComputeNodeId.removeNode(computeNodeId);
+        if (!map.tracksSomeComputeNode()) {
+            resourceIsolationGroupToTabletMapping.remove(resourceIsolationGroup);
+        }
+    }
+
+    public void modifyComputeNode(Long computeNodeId,
+                                  String oldResourceIsolationGroup, String newResourceIsolationGroup) {
+        writeLock.lock();
```

Comment:
before acquiring lock, check if old equals to new resource isolation group, and simply return if so?

Reply:
done